Docker build & publish workflow is broken

**Describe the bug**

The workflow is failing with the following error messages, in some case:

[429 error](https://github.com/penumbra-zone/penumbra/actions/runs/3514254082/jobs/5888571575)

```
error: failed to solve: busybox:1.34.1-musl: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/busybox/manifests/sha256:0497d8cd241e7a2c44564a33dddd6b1a06fe000fb36ff2b3cdc12337a0ae45f2: 429 Too Many Requests - Server message: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
Error: buildx failed with: error: failed to solve: busybox:1.34.1-musl: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/busybox/manifests/sha256:0497d8cd241e7a2c44564a33dddd6b1a06fe000fb36ff2b3cdc12337a0ae45f2: 429 Too Many Requests - Server message: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```

or [503 error](https://github.com/penumbra-zone/penumbra/actions/runs/3515449221/jobs/5890726639):

```
error: failed to solve: failed to compute cache key: failed to copy: httpReadSeeker: failed open: unexpected status code https://ghcr.io/v2/strangelove-ventures/infra-toolkit/blobs/sha256:14bccb3a1cff3b5f09bf1f21ba9ecb17c96329fb82f96605d922cfa4c9a82032: 503 first byte timeout
Error: buildx failed with: error: failed to solve: failed to compute cache key: failed to copy: httpReadSeeker: failed open: unexpected status code https://ghcr.io/v2/strangelove-ventures/infra-toolkit/blobs/sha256:14bccb3a1cff3b5f09bf1f21ba9ecb17c96329fb82f96605d922cfa4c9a82032: 503 first byte timeout
```

and [others](https://github.com/penumbra-zone/penumbra/actions/runs/3516748177/jobs/5893703023):

```
Error: buildx failed with: error: failed to solve: process "/bin/sh -c if [ \"$TARGETARCH\" = \"arm64\" ] && [ \"$BUILDARCH\" != \"arm64\" ]; then       export CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER=aarch64-linux-gnu-gcc         CC_aarch64_unknown_linux_gnu=aarch64-linux-gnu-gcc         CXX_aarch64_unknown_linux_gnu=aarch64-linux-gnu-g++         PKG_CONFIG_SYSROOT_DIR=/usr/aarch64-linux-gnu;       cargo build --release --target aarch64-unknown-linux-gnu;     elif [ \"$TARGETARCH\" = \"amd64\" ] && [ \"$BUILDARCH\" != \"amd64\" ]; then       export CARGO_TARGET_x86_64_UNKNOWN_LINUX_GNU_LINKER=x86_64-linux-gnu-gcc         CC_x86_64_unknown_linux_gnu=x86_64-linux-gnu-gcc         CXX_x86_64_unknown_linux_gnu=x86_64-linux-gnu-g++         PKG_CONFIG_SYSROOT_DIR=/usr/x86_64-linux-gnu;       cargo build --release --target x86_64-unknown-linux-gnu;     else       cargo build --release --target $(uname -m)-unknown-linux-gnu;    fi;" did not complete successfully: exit code: 101
```

Here's another example of a [build failure, from the smoketest job](https://github.com/penumbra-zone/penumbra/actions/runs/3516680306/jobs/5893547650). 

<details>
<summary>relevant error output</summary>

```
Run timeout --preserve-status $TESTNET_RUNTIME docker-compose up --exit-code-from pd-node0
using --exit-code-from implies --abort-on-container-exit
Creating network "penumbra_localnet" with driver "bridge"
Creating network "penumbra_monitoring" with the default driver
Creating volume "penumbra_prometheus_data" with default driver
Pulling tendermint-node0 (tendermint/tendermint:v0.34.21)...
v0.34.21: Pulling from tendermint/tendermint
Digest: sha256:1bfeb39391b231f99a0068ed93777e648367ebd18831f521cf8c384768fd8562
Status: Downloaded newer image for tendermint/tendermint:v0.34.21
Pulling prometheus (prom/prometheus:latest)...
latest: Pulling from prom/prometheus
toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
Error: Process completed with exit code 1.
```
</details>

--- a/.github/workflows/docker-publish.yaml
+++ b/.github/workflows/docker-publish.yaml
@@ -1,11 +1,14 @@
 name: Create and publish a Docker image
 
+# Build strategy: we want to build new containers
+# on merges into main branch (for testnet-preview)
+# and every tag (weekly testnet iterations).
 on:
   push:
     tags:
     - '**'
     branches:
-    - '**'
+    - main
 
 jobs:
   build-and-push-image:
@@ -24,12 +27,18 @@ jobs:
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v2
 
-      - name: Log in to the Container registry
+      - name: Log in to the GitHub container registry (for pushes)
         uses: docker/login-action@v2
         with:
           registry: ghcr.io
           username: ${{ github.actor }}
           password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Log in to the Docker Hub container registry (for pulls)
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
 
       - name: Extract metadata (tags, labels) for Docker
         id: meta

--- a/.github/workflows/smoke.yml
+++ b/.github/workflows/smoke.yml
@@ -15,7 +15,13 @@ jobs:
         with:
           profile: minimal
           toolchain: stable
-            
+
+      - name: Log in to the Docker Hub container registry (for pulls)
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
       - name: Build the testnet.
         run: |
           ./scripts/docker_compose_freshstart.sh
@@ -30,7 +36,7 @@ jobs:
       - name: Now start testnet in the background so we can run integration tests.
         run: |
           docker-compose up --detach
-          
+
       # DOC: please update the devnet guide if you change this (see "How to run smoke-tests")
       - name: Run integration tests against localhost.
         run: cargo test --features nct-divergence-check --package pcli -- --ignored --test-threads 1 --nocapture

